vlasiator.cpp

```diff
@@ -942,35 +944,44 @@ int simulate(int argn,char* args[]) {
                doNow[0] = 2; // Setting to 2 so as to not increment the restart count below.
                globalflags::writeRestart = false; // This flag is only used by MASTER_RANK here and it needs to be reset after a restart write has been issued.
             }
-         }
-         else {
+         } else {
             doNow[0] = 0;
          }
-         if (globalflags::balanceLoad || globalflags::doRefine) {
+         if (  P::saveRecoverTstepInterval > 0
+            && ((P::tstep % P::saveRecoverTstepInterval == 0
+                  && P::tstep != P::tstep_min)
+               || globalflags::writeRecover)
+         ) {
             doNow[1] = 1;
+            if (globalflags::writeRecover == true) {
+               globalflags::writeRecover = false; // This flag is only used by MASTER_RANK here and it needs to be reset after a recover write has been issued.
+            }
+         } else {
+            doNow[1] = 0;
+         }
+         if (globalflags::balanceLoad || globalflags::doRefine) {
+            doNow[2] = 1;
             globalflags::balanceLoad = false;
             if (globalflags::doRefine) {
-               doNow[2] = 1;
+               doNow[3] = 1;
                globalflags::doRefine = false;
             }
          }
       }
-      MPI_Bcast( &doNow, 3 , MPI_INT , MASTER_RANK ,MPI_COMM_WORLD);
-      writeRestartNow = doNow[0];
-      doNow[0] = 0;
-      if (doNow[1] == 1) {
+      MPI_Bcast( &doNow, 4 , MPI_INT , MASTER_RANK ,MPI_COMM_WORLD);
+      if (doNow[2] == 1) {
          P::prepareForRebalance = true;
-         doNow[1] = 0;
+         doNow[2] = 0;
```

Comment:
then, remove resetting `doNow` flags here in the if-clauses. (also there's a mix of `0` and `false`...)

Instead, at the end of these checks at 1060 or so, set all flags again to zero.

Reply:
*Someone* decided to add the 4th as a bool instead and is still passing it in the MPI broadcast as an MPI_INT. I disapprove but that was merged a while back.